File: src/test/java/de/kevcodez/ecommerce/parser/impl/BonprixParserTest.java
package de.kevcodez.ecommerce.parser.impl;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.AssertionsForClassTypes.withinPercentage;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

import java.math.BigDecimal;
import java.util.List;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import de.kevcodez.ecommerce.parser.domain.image.ImageDto;
import de.kevcodez.ecommerce.parser.domain.image.ImageVariant;
import de.kevcodez.ecommerce.parser.domain.price.Discount;
import de.kevcodez.ecommerce.parser.domain.price.Price;
import de.kevcodez.ecommerce.parser.domain.product.Product;

class BonprixParserTest extends AbstractParserTest {

    private static final String SAMPLE_REGULAR = "/bonprix/sample_regular";
    private static final String SAMPLE_DISCOUNTED = "/bonprix/sample_discounted";

    private static final String VALID_URL = "https://www.bonprix.de/produkt/jeans-straight-blau-905358/";

    private BonPrixParser bonPrixParser;

    @BeforeEach
    void setup() {
        bonPrixParser = new BonPrixParser(websiteSourceDownloader);
    }

    @Test
    void baseDataRegular() {
        when(websiteSourceDownloader.download(anyString())).thenReturn(getResourceAsString(SAMPLE_REGULAR));

        Product product = bonPrixParser.parse(VALID_URL);

        // Base data
        assertAll("Base Data",
            () -> assertThat(product.getTitle()).isEqualTo("Jeans Regular Fit Straight"),
            () -> assertThat(product.getUrl()).isEqualTo(VALID_URL),
            () -> assertThat(product.getExternalId()).isEqualTo("90535895"),
            () -> assertThat(product.getDescription()).startsWith("Diese Herren Jeans Regular Fit von John Baner")
        );

        assertThat(product.getPrice().getCurrentPrice()).isCloseTo(new BigDecimal("19.99"), withinPercentage(0.1D));
    }

    @Test
    void imagesRegular() {
        when(websiteSourceDownloader.download(anyString())).thenReturn(getResourceAsString(SAMPLE_REGULAR));

        Product product = bonPrixParser.parse(VALID_URL);

        List<ImageDto> images = product.getImages();
        assertThat(images).hasSize(7);

        ImageDto firstImage = images.get(0);

        assertThat(firstImage.getVariants()).hasSize(3);

        assertThat(firstImage.getVariants()).contains(ImageVariant.builder()
            .url("https://image01.bonprix.de/assets/319x448/1511338838/15045955-hDEd0CBM.jpg")
            .height(319)
            .width(448)
            .build());
        assertThat(firstImage.getVariants()).contains(ImageVariant.builder()
            .url("https://image01.bonprix.de/assets/31x44/1511338838/15045955-hDEd0CBM.jpg")
            .height(31)
            .width(44)
            .build());
        assertThat(firstImage.getVariants()).contains(ImageVariant.builder()
            .url("https://image01.bonprix.de/assets/957x1344/1511338838/15045955-hDEd0CBM.jpg")
            .height(957)
            .width(1344)
            .build());
    }

    @Test
    void sampleDiscount() {
        when(websiteSourceDownloader.download(anyString())).thenReturn(getResourceAsString(SAMPLE_DISCOUNTED));

        Product product = bonPrixParser.parse(VALID_URL);
        Price price = product.getPrice();

        assertThat(price.getCurrentPrice()).isCloseTo(new BigDecimal("9.99"), withinPercentage(0.1D));

        Discount discount = price.getDiscount();
        Discount expectedDiscount = new Discount()
            .setDiscount(BigDecimal.valueOf(3L))
            .setOldPrice(new BigDecimal("12.99"))
            .setPercentage(BigDecimal.valueOf(23.09D));

        verifyDiscount(discount, expectedDiscount);
    }

}

File: src/test/kotlin/de/kevcodez/ecommerce/parser/impl/BonprixParserTest.kt
package de.kevcodez.ecommerce.parser.impl

import com.nhaarman.mockito_kotlin.whenever
import de.kevcodez.ecommerce.parser.domain.image.ImageVariant
import de.kevcodez.ecommerce.parser.domain.price.Discount
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.AssertionsForClassTypes.withinPercentage
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.anyString
import java.math.BigDecimal

internal class BonprixParserTest : AbstractParserTest() {

    private lateinit var bonPrixParser: BonPrixParser

    @BeforeEach
    fun setup() {
        bonPrixParser = BonPrixParser(websiteSourceDownloader)
    }

    @Test
    fun baseDataRegular() {
        whenever(websiteSourceDownloader.download(anyString())).thenReturn(getResourceAsString(SAMPLE_REGULAR))

        val product = bonPrixParser.parse(VALID_URL)

        // Base data
        assertThat(product.title).isEqualTo("Jeans Regular Fit Straight")
        assertThat(product.url).isEqualTo(VALID_URL)
        assertThat(product.externalId).isEqualTo("90535895")
        assertThat(product.description).startsWith("Diese Herren Jeans Regular Fit von John Baner")

        assertThat(product.price!!.currentPrice).isCloseTo(BigDecimal("19.99"), withinPercentage(0.1))
    }

    @Test
    fun imagesRegular() {
        whenever(websiteSourceDownloader.download(anyString())).thenReturn(getResourceAsString(SAMPLE_REGULAR))

        val product = bonPrixParser.parse(VALID_URL)

        val images = product.images
        assertThat(images).hasSize(7)

        val firstImage = images[0]

        assertThat(firstImage.variants).hasSize(3)

        assertThat(firstImage.variants).contains(ImageVariant(
                url = "https://image01.bonprix.de/assets/319x448/1511338838/15045955-hDEd0CBM.jpg",
                height = 319,
                width = 448))

        assertThat(firstImage.variants).contains(ImageVariant(
                url = "https://image01.bonprix.de/assets/31x44/1511338838/15045955-hDEd0CBM.jpg",
                height = 31,
                width = 44))

        assertThat(firstImage.variants).contains(ImageVariant(
                url = "https://image01.bonprix.de/assets/957x1344/1511338838/15045955-hDEd0CBM.jpg",
                height = 957,
                width = 1344))
    }

    @Test
    fun sampleDiscount() {
        whenever(websiteSourceDownloader.download(anyString())).thenReturn(getResourceAsString(SAMPLE_DISCOUNTED))

        val product = bonPrixParser.parse(VALID_URL)
        val price = product.price!!

        assertThat(price.currentPrice).isCloseTo(BigDecimal("9.99"), withinPercentage(0.1))

        val discount = price.discount!!
        val expectedDiscount = Discount(discount = BigDecimal.valueOf(3L), oldPrice = BigDecimal("12.99"), percentage = BigDecimal.valueOf(23.09))

        verifyDiscount(discount, expectedDiscount)
    }

    companion object {

        private val SAMPLE_REGULAR = "/bonprix/sample_regular"
        private val SAMPLE_DISCOUNTED = "/bonprix/sample_discounted"

        private val VALID_URL = "https://www.bonprix.de/produkt/jeans-straight-blau-905358/"
    }

}
